services/cargohold/test/integration/API.hs

```diff
@@ -231,3 +244,82 @@ testUploadCompatibility c = do
       \test\r\n\
       \--FrontierIyj6RcVrqMcxNtMEWPsNpuPm325QsvWQ--\r\n\
       \\r\n"
+
+--------------------------------------------------------------------------------
+-- Federation behaviour
+
+testRemoteDownloadWrongDomain :: TestM ()
+testRemoteDownloadWrongDomain = do
+  assetId <- liftIO $ Id <$> nextRandom
+  uid <- liftIO $ Id <$> nextRandom
+
+  let key = AssetKeyV3 assetId AssetPersistent
+      qkey = Qualified key (Domain "invalid.example.com")
+  downloadAsset uid qkey () !!! do
+    const 422 === statusCode
+
+testRemoteDownloadNoAsset :: TestM ()
+testRemoteDownloadNoAsset = do
+  assetId <- liftIO $ Id <$> nextRandom
+  uid <- liftIO $ Id <$> nextRandom
+  let key = AssetKeyV3 assetId AssetPersistent
+      qkey = Qualified key (Domain "faraway.example.com")
+      respond req
+        | frRPC req == "get-asset" =
+          pure ("application" // "json", Aeson.encode (GetAssetResponse False))
+        | otherwise =
+          throw
+            . MockErrorResponse HTTP.status404
+            . LText.decodeUtf8With Text.lenientDecode
+            . Aeson.encode
+            $ assetNotFound
+  (_, reqs) <- withMockFederator respond $ do
+    downloadAsset uid qkey () !!! do
+      const 404 === statusCode
```

Comment:
I am a bit puzzled here. Based on the rest of the function, I'd expect the asserted status code to be 200 or such: the request's path is `"get-asset"`, and the `respond` function, in the case when that is the path, returns `pure ("application" // "json", Aeson.encode (GetAssetResponse False))`. Or does `GetAssetResponse False` map to the 404 status code? On the other hand, I understand there is no domain "faraway.example.com" running Cargohold, so it can't succeed.

Reply:
> On the other hand, I understand there is no domain "faraway.example.com" running Cargohold, so it can't succeed.

I'm not sure what you mean. We are mocking federator here, we can make any domain exist.

The 404 status code does not come from federator (which indeed returns 200), but from the public API endpoint which we are calling ('/assets/v4'), which sees that the asset does not exist, and returns 404 to the client.

Reply:
Ok, got it. Thanks for the clarification!